### ISSUE TITLE
🦠 Model Request: TOX21 NR-Aromatase

### ISSUE THREAD

**Issue by GemmaTuron** [ersilia-os/ersilia#441]
### Model Title

Tox21 NR-Aromatase (TDCommons Dataset)

### Publication

Hello @Chigoziee!

As part of your Outreachy contribution, we have assigned you the dataset "Tox21 NR-Aromatase" from the Therapeutics Data Commons to try and build a binary classification ML model.
Please copy the provided Google Colab template and use this issue to provide updates on the progress. We'll value not only being able to build the model but also interpreting its results.

**Comment by Chigoziee** [comment]
[Google colab link] (https://colab.research.google.com/drive/1eNsH2054XueZOPnnlolgdZIpvbou_FOn?usp=sharing)

I was able to download, install and load all packages and dataset successfully. I tried out some of the classifiers I found for binary classification at [Lazy-qsar](https://github.com/ersilia-os/lazy-qsar/tree/main/lazyqsar/binary).

I tested MorganBinaryClassifier. this gave me an accuracy of 75%, and after computing the Confusion matrix it gave me an F1 score of 0.14 all on the validation data.
I tested my data on the MordredBinaryClassifier. this gave me an accuracy of 76%, and after computing the confusion matrix, I calculated the F1 score to be 0, since there where no true positives in my confusion matrix of the validation data.
I finally tested on the ClassicBinaryClassifier. this gave an accuracy of 84% and an f1 score of 0.323. Having the highest F1 score from the validation data I decided to build my model using the ClassicBinaryClassifier.

I noticed my model had a very low precision (0.19) on the validation data, and had issues classifying the positives (1) in my dataset. I think this is because during training the model isn't thought well how to classify positive (1) labels because of the very skewed data. from my training data, the positive labels just accounts for only 5% of the data. this i think made the model a bit biased.

is there any way I can make my model better at predicting positive labels, with respect to the skewed dataset?

**Comment by GemmaTuron** [comment]
Hi @Chigoziee ,

There are a number of things you could try, starting by incrementing the training time of the model, and then maybe subsampling to ameliorate the big imbalance in your data.
I'll leave the issue open in case you want to add anything to this thread, when you are ready please close it and link it as a contribution to Outreachy, and go onto preparing your final application

**Comment by Femme-js** [comment]
Hi @Chigoziee ! For skewed data, you can look at my issue which can help you understand the problem. The basic idea is to get the entire data into a limited range and normalising it. 

Link to my issue: https://github.com/ersilia-os/ersilia/issues/423

**Comment by Chigoziee** [comment]
Thanks @Femme-js  I'll check it out

**Comment by Chigoziee** [comment]
@GemmaTuron after subsampling I got much better results, I tested the model it with a subsampled dataset of size 1500 molecules, 2000 molecules, 3000 molecules and 3500 molecules.

These are the results of the subsampling test:
1500 molecules gave me an ROC of 0.81 and an F1 score of 0.16 
2000 molecules gave me an ROC of 0.83 and an F1 score of 0.32
3000 molecules gave me an ROC of 0.84 and an F1 score of 0.41
3500 molecules gave me an ROC of 0.85 and an F1 score of 0 (no true positives) 

I then decided to build my model using the 3000 subsampled dataset since it had a very good ROC and also a very good F1 score.

**Comment by Chigoziee** [comment]
**Summary of NR-Aromatase model**
[Google Colab link](https://colab.research.google.com/drive/1eNsH2054XueZOPnnlolgdZIpvbou_FOn?usp=sharing)

**Major steps in building model:**

- Installation of PyTDC
- Loading the data from Therapeutics Data Commons
- Data analysis, processing and visualization
- installation of  LazyQSAR
- Testing the data with different classifiers
- Model Evaluation on Test Data and Validation Data

**Data Analysis and Visualization**

- I divided the loaded data between train/test/validation sets which are already split through the PyTDC module
- I then subsampled the dataset to a ratio of 216:2784 after noticing the original dataset was too skewed after having a ratio of 216:4075 for both classes. Summary of my dataset:
       Training data has 3000 molecules
       Validation data has 582 molecules
       Test data has 1164 molecules
- I then viewed my dataset on a pie chart using the Matplotlib library.
![data pie chart](https://user-images.githubusercontent.com/61709944/199107079-70d9783a-dd00-4280-b752-c7aedab007c4.png)
- I then divided my data set into features and labels

**Testing the data with different classifiers**

- I tested MorganBinaryClassifier. this gave me an area under a receiver operating characteristic (ROC) of 0.75, and after computing the Confusion matrix it gave me an F1 score of 0.14 all on the validation data.
- I tested my data on the MordredBinaryClassifier. this gave me an area under a receiver operating characteristic (ROC) of 0.76, and after computing the confusion matrix, I calculated the F1 score to be 0, since there where no true positives in my confusion matrix of the validation data.
- I finally tested on the ClassicBinaryClassifier. this gave an area under a receiver operating characteristic (ROC) of 0.84 and an f1 score of 0.323. Having the highest F1 score from the validation data I decided to build my model using the ClassicBinaryClassifier.

**Model Evaluation on Test Data and Validation Data**
Before subsampling I tested MorganBinaryClassifier. this gave me an area under a receiver operating characteristic (ROC) of 0.75, and after computing the Confusion matrix it gave me an F1 score of 0.14 all on the validation data. I tested my data on the MordredBinaryClassifier. this gave me an area under a receiver operating characteristic (ROC) of 0.76, and after computing the confusion matrix, I calculated the F1 score to be 0, since there where no true positives in my confusion matrix of the validation data. I finally tested on the ClassicBinaryClassifier. this gave anarea under a receiver operating characteristic (ROC) of 0.84 and an f1 score of 0.323. Having the highest F1 score from the validation data I decided to build my model using the ClassicBinaryClassifier.

Noticing that my model was not good at predicting the positive class I tried downsampling from the advice of my mentor.  
 I tested the model it with a subsampled dataset of size 1500 molecules, 2000 molecules, 3000 molecules and 3500 molecules.

These are the results of the subsampling test:
1500 molecules gave me an ROC of 0.81 and an F1 score of 0.16
2000 molecules gave me an ROC of 0.83 and an F1 score of 0.32
3000 molecules gave me an ROC of 0.84 and an F1 score of 0.41
3500 molecules gave me an ROC of 0.85 and an F1 score of 0 (no true positives)

I then decided to build my model using the 3000 subsampled dataset since it had a very good ROC and also a very good F1 score.

After subsampling I also tested the dataset on the three different classifiers I used earlier (MorganBinaryClassifier, MordredBinaryClassifier, ClassicBinaryClassifier). The ClassicBinaryClassifier still gave the best results which are the ones stated above.
the image of my ROC curve is sown below:
![ROC curve](https://user-images.githubusercontent.com/61709944/199109312-120b31b6-c9a9-4c2d-8cba-0a09d10e1cca.png)

**Comment by Chigoziee** [comment]
Hi @GemmaTuron, I don't seem to have access to closing the issue. can you please assist me with closing it. Thank you

**Comment by GemmaTuron** [comment]
Hello @Chigoziee  !

Thanks for your contributions to Ersilia during the Outreachy Application Period! We hope you have learnt and enjoyed as much as the Ersilia team did!
I will close this issue as this was part of the application period solely.

**Comment by Chigoziee** [comment]
@GemmaTuron I really did learn alot, thanks for the experience.
